mmdet3d/models/segmentors/encoder_decoder.py

```diff
@@ -0,0 +1,344 @@
+import numpy as np
+import torch
+from torch.nn import functional as F
+
+from mmseg.models import SEGMENTORS, EncoderDecoder
+from ..builder import build_backbone, build_head, build_neck
+from .base import Base3DSegmentor
+
+
+@SEGMENTORS.register_module()
+class EncoderDecoder3D(Base3DSegmentor, EncoderDecoder):
+    """3D Encoder Decoder segmentors.
+
+    EncoderDecoder typically consists of backbone, decode_head, auxiliary_head.
+    Note that auxiliary_head is only used for deep supervision during training,
+    which could be dumped during inference.
+    """
+
+    def __init__(self,
+                 backbone,
+                 decode_head,
+                 neck=None,
+                 auxiliary_head=None,
+                 train_cfg=None,
+                 test_cfg=None,
+                 pretrained=None):
+        Base3DSegmentor.__init__(self)
+        self.backbone = build_backbone(backbone)
+        if neck is not None:
+            self.neck = build_neck(neck)
+        self._init_decode_head(decode_head)
+        self._init_auxiliary_head(auxiliary_head)
+
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+
+        self.init_weights(pretrained=pretrained)
+
+        assert self.with_decode_head, \
+            '3D EncoderDecoder Segmentor should have a decode_head'
+
+    def _init_decode_head(self, decode_head):
+        """Initialize ``decode_head``"""
+        self.decode_head = build_head(decode_head)
+        self.num_classes = self.decode_head.num_classes
+
+    def encode_decode(self, points, img_metas):
+        """Encode points with backbone and decode into a semantic segmentation
+        map of the same size as input.
+
+        Args:
+            points (torch.Tensor): Input points of shape [B, N, 3+C].
+            img_metas (list[dict]): Meta information of each sample.
+
+        Returns:
+            torch.Tensor: Segmentation logits of shape [B, num_classes, N].
+        """
+        x = self.extract_feat(points)
+        out = self._decode_head_forward_test(x, img_metas)
+        return out
+
+    def forward_train(self, points, img_metas, pts_semantic_mask):
+        """Forward function for training."""
+        points_cat = torch.stack(points)
+        pts_semantic_mask_cat = torch.stack(pts_semantic_mask)
+        return EncoderDecoder.forward_train(
+            self,
+            img=points_cat,
```

Comment:
This is a little bit ugly, can we implement a more complete base class and fully based on that class?